src/client/ws/client.rs

```diff
@@ -257,164 +255,228 @@ impl<Notif> Drop for Subscription<Notif> {
+							if let Err(e) = send_back_sink.try_send(notif.params.result) {
+								log::error!("Dropping subscription {:?} error: {:?}", sub_id, e);
+								manager.remove_active_subscription(request_id, sub_id).expect("subscription is active; checked above");
+							}
+						}
+						None => {
+							log::error!("Subscription ID: {:?} not an active subscription", sub_id);
+						},
 					}
 				}
-			}
-
-			// Request for the server to unsubscribe to us has succeeded.
-			Either::Right(Ok(RawClientEvent::Unsubscribed { request_id: _ })) => {}
-
-			Either::Right(Err(RawClientError::Inner(WsConnectError::Ws(SokettoError::UnexpectedOpCode(e))))) => {
-				log::error!(
-					"Client Error: {:?}, <https://github.com/paritytech/jsonrpsee/issues/154>",
-					SokettoError::UnexpectedOpCode(e)
-				);
-			}
-			Either::Right(Err(e)) => {
-				// TODO: https://github.com/paritytech/jsonrpsee/issues/67
-				log::error!("Client Error: {:?} terminating connection", e);
-				break;
-			}
+				Some(Err(e)) => {
+					log::error!("Error: {:?} terminating client", e);
+					return;
+				}
+			},
 		}
 	}
 }
 
-/// Close subscription in RawClient helper.
-/// Logs if the subscription couldn't be found.
-async fn close_subscription(client: &mut RawClient, request_id: RawClientRequestId, unsubscribe_method: String) {
-	match client.subscription_by_id(request_id).and_then(|s| s.into_active()) {
-		Some(mut sub) => {
-			if let Err(e) = sub.close(&unsubscribe_method).await {
-				log::error!("RequestID : {:?}, unsubscribe to {} failed: {:?}", request_id, unsubscribe_method, e);
+/// Process a response from the server.
+///
+/// Returns `Ok(_)` if the response was successful or if the error could be handled.
+/// Returns `Err(_)` if the response couldn't be handled.
+fn process_response(
+	manager: &mut RequestManager,
+	response: jsonrpc::Output,
+	subscription_capacity: usize,
+) -> Result<Option<(String, jsonrpc::Params)>, Error> {
+	let response_id = match response.id().as_number() {
+		Some(n) => *n,
+		None => return Err(Error::InvalidRequestId),
+	};
+
+	match manager.request_status(&response_id) {
+		RequestStatus::PendingMethodCall => {
+			let send_back_oneshot = match manager.complete_pending_method_call(response_id) {
+				Some(send_back) => send_back,
+				None => return Err(Error::InvalidRequestId),
+			};
+			let response: Result<JsonValue, Error> = response.try_into().map_err(Error::Request);
+			match send_back_oneshot.send(response) {
+				Err(Err(e)) => Err(e),
+				_ => Ok(None),
+			}
+		}
+		RequestStatus::PendingSubscription => {
+			let (send_back_oneshot, unsubscribe_method) = match manager.complete_pending_subscription(response_id) {
+				Some(pending) => pending,
+				None => return Err(Error::InvalidRequestId),
+			};
+
+			let json_sub_id: JsonValue = match response.try_into() {
+				Ok(response) => response,
+				Err(e) => {
+					return match send_back_oneshot.send(Err(Error::Request(e))) {
+						Err(Err(e)) => Err(e),
```

Comment:
These erorr-in-errors feel a bit boilerplatey; maybe we can try to clean it up in a diff PR?

Reply:
The problem is that send returns `Result<(), T>` then T is `Result<JsonValue, Error>`, 

So, Err<Ok(_)> should be unreachable

Reply:
> Maybe explicit handling the code?

Not sure what you mean, can you elaborate?

Reply:
Yeah,  I think I got a little confused by myself for a while.

What I meant is that before I did:

```rust
				match send_back_oneshot.send(Err(Error::InvalidSubscriptionId)) {
					Err(Err(e)) => Err(e),
                    // Err(Ok(_)) -> will become Ok(None) handled by wildcard but it should be an error because it can't be handled when the channel is down.
					_ => Ok(None),
				}
```

I changed it to this:

```rust
				match send_back_oneshot.send(Err(Error::InvalidSubscriptionId)) {
					Err(Err(e)) => Err(e),
					Err(Ok(_)) => unreachable!("Error sent above; qed"),
					_ => Ok(None),
				}
```